vulnfeeds/cmd/download-cves/main.go

```diff
@@ -74,26 +83,64 @@ func downloadCVE2(apiKey string, cvePath string) {
 	if err != nil {
 		log.Fatalf("Error reading response body: %+v", err)
 	}
-	var NVD2Data cves.NVDCVE2
+	var NVD2Data *cves.NVDCVE2
 	err = json.Unmarshal(body, &NVD2Data)
 	if err != nil {
 		log.Fatalf("Failed to decode NVD data: %+v", err)
 	}
-	Logger.Infof("%d results to download", NVD2Data.TotalResults)
-	// TODO: figure out how to paginate
-	file.Write(body)
+	Logger.Infof("At offset %d of total %d results", *NVD2Data.StartIndex, *NVD2Data.TotalResults)
+	return *NVD2Data
+}
+
+// Download all of the CVE data using the 2.0 API
+// See https://nvd.nist.gov/developers/vulnerabilities
+func downloadCVE2(APIKey string, CVEPath string) {
+	file, err := os.OpenFile(path.Join(CVEPath, "nvdcve-2.0.json.new"), os.O_CREATE|os.O_RDWR|os.O_TRUNC, 0644)
+	defer file.Close()
+	if err != nil { // There's an existing file, check if it matches server file
+		Logger.Fatalf("Something went wrong when creating/opening file: %+v", err)
+	}
+	completeNVDdataset := cves.NVDCVE2{}
+	page := cves.NVDCVE2{}
+	offset := 0
+	for {
+		page = downloadCVE2WithOffset(APIKey, offset)
+		if len(page.Vulnerabilities) == 0 {
+			break
+		}
+		completeNVDdataset.Vulnerabilities = append(completeNVDdataset.Vulnerabilities, page.Vulnerabilities...)
+		offset += 2000
+		time.Sleep(6)
+	}
+	// completeNVDdataset starts out completely uninitialised
```

Comment:
We can simplify this by changing the type of `completeNVDdataset` to an array of vulnerabilities. So code from 115 to 127 can just consist of setting `page.Vulnerabilities = completeNVDdataset`, and calling `page.ToJSON(file)` instead.

Reply:
I'm not quite following, sorry.

Every page returned from the API call (and currently by `downloadCVE2WithOffset`) is a `cves.NVDCVE2` struct. Some of the metadata fields change with each subsequent page (i.e. the `StartIndex`, potentially the `TotalResults` if things are updated between calls). Definitely the `timestamp` changes.

The current approach disregards those values for the intermediate pages, only focusing on what we wind up with   after we've reached the end, because the final values are all that really matter.

I don't want each page's metadata winding up in the output, that way what gets emitted looks exactly like one page's worth of output, it's just one very big page...

Reply:
Yes, that makes sense. In the current code you are already replacing page every loop, so when the final loop finishes, the page variable will only contain the metadata for the very last page, (plus all the vulnerabilities for the final page in `page.Vulnerabilities). 

Now if we replace `page.Vulnerabilities` with the `completeNVDDataset`, which contains all the vulnerabilities concatenated together, you will have the final page's metadata, plus one very big page of vulnerabilities, similarly to what you are doing right now. 

Put another way, what is happening right now is you are filling up `completeNVDDataset.Vulnerabilities`, leaving everything else empty, then copying everything from `page` **except** the `Vulnerabilities` field over to `completeNVDDataset`. What I am suggesting is we can achieve the exact same result, by copying in the other direction, i.e. copying `completeNVDDataset.Vulnerabilties` into `page.Vulnerabilities`, and calling toJSON on `page` instead.

Reply:
Ah, thanks for clarifying, yes, that makes much more sense and reduces code. Good thinking!

Changed in https://github.com/google/osv.dev/pull/773/commits/ba10e2ab0c1677288177a158a6f59b01a5a0828a